rasa/server.py

```diff
@@ -0,0 +1,786 @@
+        except Exception as e:
+            logger.debug(traceback.format_exc())
+            raise ErrorResponse(
+                500, "ServerError", "An unexpected error occurred. Error: {}".format(e)
+            )
+
+    @app.post("/model/train")
+    @requires_auth(app, auth_token)
+    async def train(request: Request):
+        """Train a Rasa Model."""
+        from rasa.train import train_async
+
+        validate_request_body(
+            request,
+            "You must provide training data in the request body in order to "
+            "train your model.",
+        )
+
+        rjs = request.json
+        validate_request(rjs)
+
+        # create a temporary directory to store config, domain and
+        # training data
+        temp_dir = tempfile.mkdtemp()
+
+        config_path = os.path.join(temp_dir, "config.yml")
+        dump_obj_as_str_to_file(config_path, rjs["config"])
+
+        if "nlu" in rjs:
+            nlu_path = os.path.join(temp_dir, "nlu.md")
+            dump_obj_as_str_to_file(nlu_path, rjs["nlu"])
+
+        if "stories" in rjs:
+            stories_path = os.path.join(temp_dir, "stories.md")
+            dump_obj_as_str_to_file(stories_path, rjs["stories"])
+
+        domain_path = DEFAULT_DOMAIN_PATH
+        if "domain" in rjs:
+            domain_path = os.path.join(temp_dir, "domain.yml")
+            dump_obj_as_str_to_file(domain_path, rjs["domain"])
+
+        try:
+            model_path = await train_async(
+                domain=domain_path,
+                config=config_path,
+                training_files=temp_dir,
+                output=rjs.get("out", DEFAULT_MODELS_PATH),
+                force_training=rjs.get("force", False),
+            )
+
+            return await response.file(model_path)
+        except Exception as e:
+            logger.debug(traceback.format_exc())
+            raise ErrorResponse(
+                500,
+                "ServerError",
+                "An unexpected error occurred during training. Error: {}".format(e),
+            )
+
+    def validate_request(rjs):
+        if "config" not in rjs:
+            raise ErrorResponse(
+                400,
+                "BadRequest",
+                "The training request is missing the required key " "`config`.",
+            )
+
+        if "nlu" not in rjs and "stories" not in rjs:
+            raise ErrorResponse(
+                400,
+                "BadRequest",
+                "To train a Rasa model you need to specify at least one type of "
+                "training data. Add `nlu` and/or `stories` to the request.",
+            )
+
+        if "stories" in rjs and "domain" not in rjs:
+            raise ErrorResponse(
+                400,
+                "BadRequest",
+                "To train a Rasa model with story training data, you also need to "
+                "specify the `domain`.",
+            )
+
+    @app.post("/model/evaluate/stories")
```

Comment:
should we also name this `test` instead of `evaluate`? or rather somewhen later? :-D

Reply:
yes let's rename it